File: JavaScript/src/warmup.js
function change(amount) {
  if (amount < 0) {
    throw new RangeError('Input must be a positive value.');
  }
  let currentValue = amount;
  const denominations = [25, 10, 5, 1];
  const answer = [];
  denominations.forEach((coin) => {
    answer.push(Math.floor(currentValue / coin));
    currentValue %= coin;
  });

  return answer;
}

function stripQuotes(str) {
  return (str.replace(/"|'/g, ''));
}

function scramble(str) {
  let scrambledString = '';
  const word = str.split('');
  while (word.length) {
    const randomIndex = Math.floor(Math.random() * word.length);
    scrambledString += word.splice(randomIndex, 1);
  }
  return scrambledString;
}

function powers(base, limit, callback) {
  let power = 0;
  let currentValue = base ** power;
  while (currentValue <= limit) {
    callback(currentValue);
    power += 1;
    currentValue = base ** power;
  }
}

function powersGenerator(base, limit) {
  let power = 0;
  return {
    next: () => {
      const currentValue = base ** power;
      const overLimit = currentValue > limit;
      power += 1;
      return {
        value: overLimit ? undefined : currentValue,
        done: overLimit,
      };
    },
  };
}

function say(input) {
  let words = '';

  const chain = (str) => {
    if (str) {
      words += str + " ";
    }

    return str ? chain : words.trim();
  };

  return chain(input);
}

function interleave(array, ...rest) {
  const shorter = array.length > rest.length ? rest : array;
  const longer = array.length > rest.length ? array : rest;
  const answer = [];

  for (let i = 0; i < shorter.length; i += 1) {
    answer.push(array[i]);
    answer.push(rest[i]);
  }

  return answer.concat(longer.slice(shorter.length, longer.length));
}

function cylinder(data) {
  let { height = 1, radius = 1 } = data;

  function surfaceArea() {
    return (Math.PI * 2 * radius * height) + (Math.PI * 2 * (radius ** 2));
  }

  function volume() {
    return Math.PI * (radius ** 2) * height;
  }

  function widen(widenFactor) {
    radius *= widenFactor;
  }

  function stretch(stretchFactor) {
    height *= stretchFactor;
  }

  function toString() {
    return `Cylinder with radius ${radius} and height ${height}`;
  }

  return Object.freeze({
    get height() {
      return height;
    },
    get radius() {
      return radius;
    },
    surfaceArea,
    volume,
    widen,
    stretch,
    toString,
  });
}

function makeCryptoFunctions(key, algorithm) {
  const crypto = require('crypto');
  return [
    function encrypt(str) {
      const cipher = crypto.createCipher(algorithm, key);
      const crypted = cipher.update(str, 'utf8', 'hex');
      return crypted + cipher.final('hex');
    },
    function decrypt(str) {
      const decipher = crypto.createDecipher(algorithm, key);
      const decrypted = decipher.update(str, 'hex', 'utf8');
      return decrypted + decipher.final('utf8');
    },
  ];
}

function randomName(data) {
  const request = require('request-promise');
  const {gender, region} = data;

  const formData = {
    gender,
    region,
    amount: 1,
  };

  let answer = '';
  const options = {
    url: 'https://uinames.com/api/',
    qs: formData,
    json: true,
    callback: (error, response, apiData) => {
      answer = (`${apiData.surname}, ${apiData.name}`);
    },
  };

  return request(options).then(() => answer);
}


module.exports = {
  change,
  stripQuotes,
  scramble,
  powers,
  powersGenerator,
  say,
  interleave,
  cylinder,
  makeCryptoFunctions,
  randomName,
};

File: Python/warmup.py
import re
from random import sample
from math import pi
from itertools import zip_longest, chain
from ast import literal_eval
from Crypto.Cipher import AES
import requests


def change(amount):
  if amount < 0:
    raise ValueError('amount cannot be negative')
  answer = []
  remaining = amount
  for coin in (25, 10, 5, 1):
    quotient, remainder = divmod(remaining, coin)
    answer.append(quotient)
    remaining = remainder
  return tuple(answer)

def strip_quotes(string):
  return re.sub('\'|\"', '', string)

def scramble(string):
  return ''.join(sample(string, len(string)))

def say(string=''):
  words = []
  def say_more(string=''):
    if string:
      words.append(string)
    return say_more if string else ' '.join(words)
  return say_more(string)

def triples(limit):
  return sorted(((a, b, c) for c in range(1, limit + 1) for b in range(1, c)
                 for a in range(1, b) if a ** 2 + b ** 2 == c ** 2))

def powers(base, limit):
  base_to_the_power = 1

  while base_to_the_power <= limit:
    yield base_to_the_power
    base_to_the_power *= base

def interleave(list1, *rest):
  zipped = zip_longest(list1, rest)
  return [element for element in chain(*zipped) if element != None]

class Cylinder:
  def __init__(self, radius=1, height=1):
    self.radius = radius
    self.height = height

  @property
  def volume(self):
    return pi * self.radius ** 2 * self.height

  @property
  def surface_area(self):
    return 2 * pi * (self.radius ** 2 + self.radius * self.height)

  def widen(self, factor):
    self.radius *= factor

  def stretch(self, factor):
    self.height *= factor

def make_crypto_functions(key, initialization_vector):
  byte_key = bytes(key, encoding='UTF-8')
  byte_IV = bytes(initialization_vector, encoding='UTF-8')

  def encrypt(byte_string):
    cipher = AES.new(byte_key, AES.MODE_CBC, byte_IV)
    return cipher.encrypt(byte_string)

  def decrypt(byte_string):
    cipher = AES.new(byte_key, AES.MODE_CBC, byte_IV)
    return cipher.decrypt(byte_string)

  return (encrypt, decrypt)

def random_name(gender=None, region=None):
  query_parameters = dict(gender=gender, region=region, amount=1)
  request = requests.get('https://uinames.com/api/', query_parameters)
  response_dict = literal_eval(request.text)
  if 'error' in response_dict.keys():
    raise ValueError(request.text)
  return response_dict['surname'] + ', ' + response_dict['name']
